rfcs/0191-worker-manager-launch-configs.md

```diff
@@ -0,0 +1,122 @@
+
+## Details
+
+New entity will be introduced to worker-manager - **Launch Configuration**.
+Each worker pool will have zero or more **launch configurations** associated with it.
+
+Each **launch configuration** will be assigned a unique ID (hash of its properties), stored in the database and will be immutable (except for the status flag).
+
+Generated ID will include the following properties:
+
+* top-level config properties (`minCapacity`, `maxCapacity`, `scalingRatio`, `lifecyle`)
+* launch configuration itself
+
+During updates of the worker pool configurations, all existing launch configurations that are not present in the new configuration will be marked as `archived`. Launch configurations that have same unique ID (hash) will be kept active. All other will be created as new launch configurations and marked as active.
+
+### Launch Configuration weight
+
+Provisioned workers would be associated with a specific launch configuration (`worker.launchConfigurationId`).
+This will allow us to to know how many workers with this launch configuration were successfully registered and claimed work.
+
+Each launch configuration will have a dynamic `weight` property that will be adjusted automatically based on the following events and metrics:
+
+* total number of successful worker provisioning attempts / registrations
+* total number of failed worker provisioning attempts / registrations
+* has any worker claimed a task
+* fail-to-success ratio over a specific time period (i.e. last hour)
+* number of non-stopped workers currently running
+
+Weight will be used to determine the likelihood of selecting a specific launch configuration for provisioning.
+
+This will be calculated at the provisioning time.
+
+### Making worker-manager extensible
+
+Worker-manager will publish additional events to Pulse to allow external systems to react:
+
+* `launch-configuration-created`
+* `launch-configuration-archived`
+* `launch-configuration-paused`
+* `launch-configuration-resumed`
+* `worker-registered`
+* `worker-failure` (provisioning or starting up failure, will include `workerPoolId` and `launchConfigurationId`)
+
+New API endpoints will be introduced:
+
+* `workerManager.getLaunchConfigs(workerPoolId)` - to retrieve all launch configurations with their statuses
+* `workerManager.getLaunchConfig(workerPoolId, launchConfigId)` - to retrieve specific launch configuration
+* `workerManager.pauseLaunchConfig(workerPoolId, launchConfigId)` - to deactivate specific active launch configuration (time bound)
+* `workerManager.resumeLaunchConfig(workerPoolId, launchConfigId)` - to resume specific active launch configuration
+
+Last two endpoints would allow external systems to pause/resume specific launch configurations based on their own criteria.
+This might be useful when dynamic adjustment of weights is not enough to prevent provisioning workers in a specific region.
```

Comment:
This is great!

In the past, Generic-worker has supported something very similar, basically a hash of its config. It would consult aws-provisioner or worker-manager and, if the hash had changed, stop taking tasks and shut down (for cloud-based workers).

I don't know if it would make sense to do the same with launch configs. In terms of handling load, it's definitely better to have a working worker keep working. But when updating a worker config, e.g., to rotate a secret or upgrade worker version, it makes sense to have old workers terminate as soon as possible.

Maybe it would make sense to support "draining" a launch config, and a worker can check for that status before claiming a task. Maybe that doesn't make sense in this RFC, but something to look forward to.

Reply:
Nice!
I was thinking that launch config should have two flags:
1. `is_archived` to mark old configs as inactive
2. `is_paused` to mark if current launch config is paused (and excluded from provisioning)

Based on this, worker would only need to know his `launch_config_id` he can use to check if it was archived yet.

I believe this would replace or extend the existing `deploymentId` logic?

Reply:
That makes a lot of sense!